src/include/gurt/common.h

```diff
@@ -329,13 +329,40 @@ d_realpath(const char *path, char *resolved_path) _dalloc_;
 	})
 
 /* Internal helper macros, not to be called directly by the outside caller */
-#define __D_PTHREAD(fn, x)						\
-	({								\
-		int _rc;						\
-		_rc = fn(x);						\
-		D_ASSERTF(_rc == 0, "%s rc=%d %s\n", #fn, _rc,		\
-			  strerror(_rc));				\
-		d_errno2der(_rc);					\
+#define __D_PTHREAD(fn, x)                                                                         \
+	({                                                                                         \
+		int _rc;                                                                           \
+		_rc = fn(x);                                                                       \
+		D_ASSERTF(_rc == 0, #fn "(%p) rc=%d %s\n", x, _rc, strerror(_rc));                 \
+		d_errno2der(_rc);                                                                  \
+	})
+
+#define __D_PTHREAD_TIMED(fn, fn2, x)                                                              \
+	({                                                                                         \
+		int _rc;                                                                           \
+		_rc = fn((x));                                                                     \
+		if (_rc == EBUSY) {                                                                \
+			int             delay1 = 0;                                                \
+			int             delay2 = 1;                                                \
+			int             f      = delay1 + delay2;                                  \
+			int             c      = f;                                                \
+			struct timespec _wait  = {};                                               \
+			clock_gettime(CLOCK_REALTIME, &_wait);                                     \
+			_wait.tv_sec += f;                                                         \
+			D_DEBUG(DB_MEM, #fn "(%p) held, waiting", x);                              \
+			while ((_rc = fn2((x), &_wait)) == ETIMEDOUT) {                            \
+				delay1 = delay2;                                                   \
+				delay2 = f;                                                        \
```

Comment:
why not just call fn2 (timed lock) right away with a 1 second timeout instead of try-locking first? is it to avoid clock_gettime() in a happy path?

Reply:
Yes, exactly, and avoid any logging overhead.  Uncontended locks should be cheap.  I hadn't thought about having just one function though and it would work just as well so I'll take a look at the code tomorrow to see if that simplifies things.